### PR TITLE
Hardcoded routing test (for sigmoid gating dendritic network)

### PR DESCRIPTION
**Pull request by karangrewal**
Here's an implementation of the hardcoded routing test where the dendritic network uses the `GatingDendriticLayer`. The context weights are hardcoded, and set using the function `get_gating_context_weights` in `frameworks/dendrites/routing/utils`.

### CODE REVIEW

File: nupic/research/frameworks/dendrites/routing/hardcoded.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    context_inds_test = randint(low=0, high=k, size=batch_size).tolist()
+    context_test = torch.stack(
+        [context_vectors[j, :] for j in context_inds_test], dim=0
+    )
+
+    target = r(context_inds_test, x_test)
+    actual = dendritic_network(x_test, context_test)
+
+    # Report mean absolute error
+    mean_abs_error = torch.abs(target - actual).mean().item()
```
Comment (subutai) [inline nupic/research/frameworks/dendrites/routing/hardcoded.py:111]: Return mean_abs_error from this function, and put the printouts in the main.

File: nupic/research/frameworks/dendrites/routing/hardcoded.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    # Run the hardcoded routing test with a dendritic network that uses "dendrites as
+    # gating"; this achieves near-zero mean absolute error, which strongly suggests
+    # that dendritic networks can successfully route
+    d_in = 100
+    d_out = 100
+    num_contexts = 10
+    d_context = 100
+
+    run_hardcoded_routing_test(
```
Comment (subutai) [inline nupic/research/frameworks/dendrites/routing/hardcoded.py:127]: Create a unit test out of this, ensure the returned mean_abs_error is within some epsilon. This way we can test that changes to our dendrites will still pass this test.

File: nupic/research/frameworks/dendrites/routing/hardcoded.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/routing/hardcoded.py:111]: Fixed.

File: nupic/research/frameworks/dendrites/routing/hardcoded.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/routing/hardcoded.py]: I added a single test case that tests whether mean_abs_error < epsilon for epsilon = 0.01.

File: nupic/research/frameworks/dendrites/dendritic_weights.py [hunk 3]
```diff
@@ -86,3 +87,22 @@ def apply_dendrites(self, y, dendrite_activations):
         """Apply dendrites as a gating mechanism."""
         # Multiple by the sigmoid of the max along each segment.
         return y * sigmoid(dendrite_activations.max(dim=2).values)
+
+
+class AbsoluateMaxGatingDendriticLayer(BiasingDendriticLayer):
```
Comment (subutai) [inline nupic/research/frameworks/dendrites/dendritic_weights.py:92]: "Absolute"

**Review by subutai — APPROVED** [review]
One spelling mistake, otherwise looks good to me!

File: nupic/research/frameworks/dendrites/dendritic_weights.py (hunk 3 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/dendritic_weights.py]: Fixed.
